[ResourceMonitor] Cannot go back to previous page by Back button after unloading of iframe.

basuke opened the pull request:
#### 1dea6a26618ba91c0e3a6dee256faa09b1e473ef
<pre>
[ResourceMonitor] Cannot go back to previous page by Back button after unloading of iframe.
<a href="https://bugs.webkit.org/show_bug.cgi?id=289623">https://bugs.webkit.org/show_bug.cgi?id=289623</a>
<a href="https://rdar.apple.com/146671361">rdar://146671361</a>

Reviewed by Brady Eidson.

When ResourceMonitor unloads iframe, a placeholder is displayed using `srcdoc`. Setting `srcdoc` is
a navigation so this action is pushed onto the history. This is not expected behavior.

Adding optional parameter to set the visibility of the navigation in the history when setting `srcdoc`.
The visibility is stored in iframe element and used when SubstituteData is created.

* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::setSrcdoc):
* Source/WebCore/html/HTMLIFrameElement.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::defaultSubstituteDataForURL):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::showResourceMonitoringError):

Canonical link: <a href="https://commits.webkit.org/292053@main">https://commits.webkit.org/292053@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6e53e9aef0300c31900c6bd0e19dd5cb56032d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94805 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14396 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4233 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99821 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45294 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96852 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14671 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22813 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72316 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29617 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97806 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10941 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85604 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52648 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10638 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3319 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44634 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3417 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101864 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21833 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81317 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22087 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81622 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80698 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20163 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25262 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2663 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15066 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21815 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26930 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21470 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24944 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23209 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1663,12 +1664,15 @@ SubstituteData FrameLoader::defaultSubstituteDataForURL(const URL& url)
+    RefPtr iframeElement = dynamicDowncast<HTMLIFrameElement>(m_frame->ownerElement());
+    ASSERT(iframeElement);
+
+    auto& srcdoc = iframeElement->attributeWithoutSynchronization(srcdocAttr);
```
Seeing pointers assumed to be non-null with debug-only assertions makes me sad.

I know it's preexisting, but - since we're expanding the use of these possibly-null objects - let's take the opportunity to also clean this up a bit.

Let's either find a way to early return or decide that RELEASE_ASSERT is safe to do.

(FrameLoader.cpp is in 3 SaferCPPExpectations files, and I'm guessing some of this wouldn't fly if it were being checked.)

basuke replied:
In this case, quick way to achieve this is replacing `ASSERT` with early return.
```c++
    RefPtr iframeElement = dynamicDowncast<HTMLIFrameElement>(m_frame->ownerElement());
    if (!iframeElement)
        return SubstituteData();
```

beidson reviewed (approved):
r+ with a change request.

basuke committed:
[ResourceMonitor] Cannot go back to previous page by Back button after unloading of iframe.
https://bugs.webkit.org/show_bug.cgi?id=289623
rdar://146671361

Reviewed by Brady Eidson.

When ResourceMonitor unloads iframe, a placeholder is displayed using `srcdoc`. Setting `srcdoc` is
a navigation so this action is pushed onto the history. This is not expected behavior.

Adding optional parameter to set the visibility of the navigation in the history when setting `srcdoc`.
The visibility is stored in iframe element and used when SubstituteData is created.

* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::setSrcdoc):
* Source/WebCore/html/HTMLIFrameElement.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::defaultSubstituteDataForURL):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::showResourceMonitoringError):

Canonical link: https://commits.webkit.org/292053@main
